### ISSUE TITLE
Removing/"forgetting" objects using configuration rather than command line

### ISSUE THREAD

**Issue by jamiekt** [hashicorp/terraform#33226]
### Terraform Version

```shell
Terraform v1.4.6
on linux_amd64
```

### Use Cases

Resources sometimes have a legitimate need to be removed from a terraform configuration but not actually removed from the deployed infrastructure. There are probably many such use cases, ours is in regard to tables in Snowflake databases.

We wish to manage our Snowflake objects using https://github.com/Snowflake-Labs/schemachange rather than terraform. The list of reasons why is varied but they can all be basically summed up as

> terraform isn't an ideal tool for managing database objects. schemachange is a tool for performing database migrations which is a popular method for managing database objects

We also have a quite a lot of issues with the Snowflake terraform provider which we regard as buggy, it doesn't handle provider upgrades very well. Hence we are migrating to schemachange to manage the majority of our Snowflake resources.

Anyway, the point is that sometimes there are legitimate reasons for resources to be removed from a terraform configuration and thus removed from terraform state. 

### Attempted Solutions

Today the only way to remove resources from terraform state is to issue `terraform state rm` commands which have to be issued prior to `terraform apply`. This isn't ideal where one is deploying using a CI pipeline because that CI pipeline has to be changed to issue these `terraform state rm` commands.

This also does not "fit" with the declarative nature of terraform configurations, it feels "hacky".

### Proposal

terraform has already solved the problem of having to issue `terraform state mv` commands using `moved` blocks:

```hcl
moved {
  from = "old.resource.address"
  to = "new.resource.address"
}
```
Terraform handles these elegantly during the planning phase and doesn't complain after the resources have been "moved" in the `apply` phase.

I propose a similar mechanism for eliminating `terraform state rm` commands ... `removed` blocks:

```hcl
removed {
  from = "resource.address"
}
```

terraform could check during `plan` (and `validate`) that there are no references to the removed resources then remove them from the state file during `terraform apply`.

**Comment by apparentlymart** [comment]
Hi @jamiekt! Thanks for sharing this use-case.

This possibility is something we've had on our radar for a long time now as we move towards replacing all of the weird manipulation commands like `state mv` and `state rm` with functionality built into the main plan and apply flow.

We have a language design sketch almost exactly like what you showed here in our internal notes, with only some debate about what exactly that block should be called -- we've also considered "forgotten" as a term that might be clearer that the real underlying object won't be deleted, whereas "removed" could be misunderstood as a synonym for "deleted".

The main challenges with a "config-driven" version of this particular operation is defining it in such a way that makes sense to appear in shared modules, as we did with `moved`, which requires carefully considering all of the different permutations of if someone upgrades over many versions of the module at once and "misses" the existence of the `removed` block, and what happens when the object is already gone when we refresh, etc. We believe these are solvable problems but we decided to do `moved` first to get the design experience and feedback before trying more "risky" operations like importing and "forgetting".

With all of this said then: this is something we do intend to do eventually and have already done some early design work for, but I can't say yet when it's likely to come since I expect we'll want to learn from the feedback from the forthcoming config-driven import (coming in v1.5, which is in beta as I'm writing this) to help inform how config-driven remove ought to work internally.

Thanks again!

**Comment by jamiekt** [comment]
@apparentlymart while it doesn’t come as a surprise that you’re already thinking about this its great to learn that indeed you are. 
i look forward to seeing this in a future terraform version.

thank you for the very quick reply.

**Comment by jamiekt** [comment]
This is also the first I’ve heard of “config driven import” which sounds intriguing.  I’m going to go and investigate that now.

**Comment by Maxwell2022** [comment]
Do you have an ETA for this feature @apparentlymart? 

Is there a work around it (except running the `terraform state rm`)?

**Comment by crw** [comment]
You can find more information about Import Blocks (config-driven import) here: https://developer.hashicorp.com/terraform/language/import

**Comment by Maxwell2022** [comment]
AFAIK import blocks are not able to remove resources from state (without deleting them)

**Comment by crw** [comment]
Thanks, I misunderstood your request for information.

**Comment by Satak** [comment]
Can't wait this to be implemented. Really useful feature. Maybe in `1.7` 🤞

**Comment by kvanzuijlen** [comment]
@apparentlymart I'm interested in working on this. Is this something that could be contributed by an external collaborator?

**Comment by jamiekt** [comment]
This feature was released yesterday in terraform 1.7 https://www.hashicorp.com/blog/terraform-1-7-adds-test-mocking-and-config-driven-remove
